### PR TITLE
[LoongArch] Support LA V1.1 feature that div.w[u] and mod.w[u] instructions with inputs not signed-extended.

### PR DESCRIPTION
**Pull request by tangaac**
Two options for clang
  -mdiv32:                 Use div.w[u] and mod.w[u] instructions with input not sign-extended.
  -mno-div32:              Do not use div.w[u] and mod.w[u] instructions with input not sign-extended.
The default is -mno-div32.

### CODE REVIEW

File: llvm/lib/Target/LoongArch/LoongArch.td [hunk 1]
```diff
@@ -118,6 +118,12 @@ def FeatureLAM_BH
                         "Support amswap[_db].{b/h} and amadd[_db].{b/h} instructions.">;
 def HasLAM_BH : Predicate<"Subtarget->hasLAM_BH()">;
 
+// Assume div.w[u] and mod.w[u] can handle inputs that are not sign-extended.
+def FeatureDiv32
+    : SubtargetFeature<"div32", "HasDiv32", "true",
+                        "Support div.w[u] and mod.w[u] can handle inputs that are not sign-extended">;
```
Comment (heiher) [inline llvm/lib/Target/LoongArch/LoongArch.td:124]: Clang is not the only LLVM frontend, and the 3A5000 is still not outdated, so I recommend setting the `div32` feature to `false` by default.

File: llvm/test/CodeGen/LoongArch/ir-instruction/sdiv-udiv-srem-urem.ll [hunk 2]
```diff
@@ -121,7 +121,7 @@ define i32 @sdiv_i32(i32 %a, i32 %b) {
 ; LA64:       # %bb.0: # %entry
 ; LA64-NEXT:    addi.w $a1, $a1, 0
 ; LA64-NEXT:    addi.w $a0, $a0, 0
-; LA64-NEXT:    div.d $a0, $a0, $a1
+; LA64-NEXT:    div.w $a0, $a0, $a1
```
Comment (heiher) [inline llvm/test/CodeGen/LoongArch/ir-instruction/sdiv-udiv-srem-urem.ll:124]: These cases are allowed when `div32` is enabled, as it eliminates unnecessary sign-extend instructions.

https://github.com/llvm/llvm-project/blob/7e85cb8a8a9de57ed10635b843662148a87b17e5/llvm/lib/Target/LoongArch/LoongArchOptWInstrs.cpp#L142-L146

File: llvm/lib/Target/LoongArch/LoongArch.td (hunk 1 above)
Comment (SixWeining) [reply llvm/lib/Target/LoongArch/LoongArch.td:124]: The `true` here is not the default value. It means when we have feature `div32`, then set `HasDiv32` to `true`.

But the test doesn't show how this feature work. I think it need some changes.

**Review by SixWeining — CHANGES_REQUESTED** [review]

**Comment by SixWeining** [comment]
The expectation is at least for this test:
```
define i32 @test(i32 %a, i32 %b) {
  %c = udiv i32 %a, %b
  ret i32 %c
}
```

### without this feature
```
        addi.w  $a1, $a1, 0
        addi.w  $a0, $a0, 0
        div.wu  $a0, $a0, $a1
        ret
```

### with this feature
```
        div.wu  $a0, $a0, $a1
        ret
```

**Comment by tangaac** [comment]
the original test is from gcc
```cpp
int divw(long a, long b) { 
  return (int)a / (int)b; 
}

unsigned int divwu(long a, long b) { 
  return (unsigned int)a / (unsigned int)b; 
}

int modw(long a, long b) { 
  return (int)a % (int)b; 
}

unsigned int modwu(long a, long b) { 
  return (unsigned int)a % (unsigned int)b; 
}
```

File: llvm/lib/Target/LoongArch/LoongArch.td (hunk 1 above)
Comment (tangaac) [reply llvm/lib/Target/LoongArch/LoongArch.td:124]: New commit has added the test.

File: llvm/lib/Target/LoongArch/LoongArchISelLowering.cpp [hunk 3]
```diff
@@ -141,7 +141,8 @@ LoongArchTargetLowering::LoongArchTargetLowering(const TargetMachine &TM,
 
     setOperationAction(ISD::BITREVERSE, MVT::i32, Custom);
     setOperationAction(ISD::BSWAP, MVT::i32, Custom);
-    setOperationAction({ISD::UDIV, ISD::UREM}, MVT::i32, Custom);
+    setOperationAction({ISD::SDIV, ISD::UDIV, ISD::SREM, ISD::UREM}, MVT::i32,
```
Comment (SixWeining) [inline llvm/lib/Target/LoongArch/LoongArchISelLowering.cpp:144]: Is this a relevant change? Or can it be split into a separate PR?

File: llvm/test/CodeGen/LoongArch/ir-instruction/sdiv-udiv-srem-urem-div32.ll [hunk 4]
```diff
@@ -0,0 +1,75 @@
+; NOTE: Assertions have been autogenerated by utils/update_llc_test_checks.py
+; RUN: llc --mtriple=loongarch64 -mattr=+d,-div32 < %s | FileCheck %s --check-prefix=LA64
+; RUN: llc --mtriple=loongarch64 -mattr=+d,+div32 < %s | FileCheck %s --check-prefix=LA64-DIV32
+
+define dso_local noundef signext i32 @divw(i64 noundef %0, i64 noundef %1) local_unnamed_addr #0 {
```
Comment (SixWeining) [inline llvm/test/CodeGen/LoongArch/ir-instruction/sdiv-udiv-srem-urem-div32.ll:5]: Remove useless attributes like `dso_local` `#0` ... See exist tests for reference.

**Comment by SixWeining** [comment]
> the original test is from gcc
> 
> ```c++
> int divw(long a, long b) { 
>   return (int)a / (int)b; 
> }
> 
> unsigned int divwu(long a, long b) { 
>   return (unsigned int)a / (unsigned int)b; 
> }
> 
> int modw(long a, long b) { 
>   return (int)a % (int)b; 
> }
> 
> unsigned int modwu(long a, long b) { 
>   return (unsigned int)a % (unsigned int)b; 
> }
> ```

Prefer to use simple IR like what I mentioned above.

File: llvm/lib/Target/LoongArch/LoongArchISelLowering.cpp (hunk 3 above)
Comment (tangaac) [reply llvm/lib/Target/LoongArch/LoongArchISelLowering.cpp:144]: This PR is related with #92205, except for the part of div32
A separate PR may be a good choice.

File: llvm/lib/Target/LoongArch/LoongArch.td (hunk 1 above)
Comment (SixWeining) [inline llvm/lib/Target/LoongArch/LoongArch.td:124]: Support -> Assume

File: llvm/test/CodeGen/LoongArch/ir-instruction/sdiv-udiv-srem-urem-div32.ll [hunk 5]
```diff
@@ -0,0 +1,75 @@
+; NOTE: Assertions have been autogenerated by utils/update_llc_test_checks.py
+; RUN: llc --mtriple=loongarch64 -mattr=+d,-div32 < %s | FileCheck %s --check-prefix=LA64
+; RUN: llc --mtriple=loongarch64 -mattr=+d,+div32 < %s | FileCheck %s --check-prefix=LA64-DIV32
+
+define i32 @divw(i64 %0, i64 %1) {
```
Comment (SixWeining) [inline llvm/test/CodeGen/LoongArch/ir-instruction/sdiv-udiv-srem-urem-div32.ll:5]: Use non-number-variable-name so that we can backport it to older llvm versions.

File: llvm/test/CodeGen/LoongArch/ir-instruction/sdiv-udiv-srem-urem-div32.ll [hunk 6]
```diff
@@ -0,0 +1,75 @@
...
+; LA64-NEXT:    addi.w $a1, $a1, 0
+; LA64-NEXT:    addi.w $a0, $a0, 0
+; LA64-NEXT:    div.w $a0, $a0, $a1
+; LA64-NEXT:    ret
+;
+; LA64-DIV32-LABEL: divw:
+; LA64-DIV32:       # %bb.0:
+; LA64-DIV32-NEXT:    div.w $a0, $a0, $a1
+; LA64-DIV32-NEXT:    ret
+  %3 = trunc i64 %0 to i32
```
Comment (SixWeining) [inline llvm/test/CodeGen/LoongArch/ir-instruction/sdiv-udiv-srem-urem-div32.ll:17]: Ditto.

**Review by SixWeining — COMMENTED** [review]
Please remove the changes of #117298.

File: llvm/test/CodeGen/LoongArch/ir-instruction/sdiv-udiv-srem-urem-div32.ll [hunk 7]
```diff
@@ -0,0 +1,79 @@
...
+
+define i32 @divw(i64 %a, i64 %b) {
+; LA64-LABEL: divw:
+; LA64:       # %bb.0:
+; LA64-NEXT:    addi.w $a0, $a0, 0
+; LA64-NEXT:    addi.w $a1, $a1, 0
+; LA64-NEXT:    div.d $a0, $a0, $a1
+; LA64-NEXT:    ret
+;
+; LA64-DIV32-LABEL: divw:
```
Comment (SixWeining) [inline llvm/test/CodeGen/LoongArch/ir-instruction/sdiv-udiv-srem-urem-div32.ll:13]: Leave a `TODO` here that the issue will be addressed by another PR.

**Review by SixWeining — APPROVED** [review]
